File: solution/067AddBinary/067AddBinary.cpp
#include <sstream>
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <unordered_map>
#include <algorithm>
#include <limits>

using namespace std;

/*
https://oj.leetcode.com/problems/add-binary/

Given two binary strings, return their sum (also a binary string).

For example,
a = "11"
b = "1"
Return "100".
*/

class Solution {
public:
    string addBinary(string a, string b) {
    	string o;
    	int m = a.size();
    	int n = b.size();
    	if(m>=n){
    		o = a;
    	}else{
    		return addBinary(b, a);
    	}  	
    	char c = 0;
    	for(int i=0; i<n; ++i){
    		c += a[m-1-i]-'0'+b[n-1-i]-'0';
    		o[m-1-i]= (c&0x1)+'0';
    		c >>= 1; 
    	}
    	for(int i=0; i<m-n; ++i){
    		c += a[m-n-1-i]-'0';
    		o[m-n-1-i] = (c&0x1)+'0';
    		c >>= 1;
    	}

    	return c==1? "1"+o: o; 
    }
};

int main(){
	Solution s;

	cout<<s.addBinary("11", "1")<<endl;

	return 0;
}

File: solution/067AddBinary/067AddBinary.go
package main

import (
	"fmt"
)

func addBinary(a string, b string) string {
	na := len(a)
	nb := len(b)
	nc := na + 1
	if na < nb {
		nc = nb + 1
	}
	r := make([]byte, nc)
	c := byte(0)
	for na != 0 && nb != 0 {
		na--
		nb--
		nc--
		d := byte(a[na]-'0') + byte(b[nb]-'0') + c
		c = d / 2
		r[nc] = (d % 2) + '0'
	}
	for na != 0 {
		na--
		nc--
		d := byte(a[na]-'0') + c
		c = d / 2
		r[nc] = (d % 2) + '0'
	}
	for nb != 0 {
		nb--
		nc--
		d := byte(b[nb]-'0') + c
		c = d / 2
		r[nc] = (d % 2) + '0'
	}
	if c != 0 {
		r[0] = c + '0'
		return string(r)
	}
	return string(r[1:])
}

func main() {
	fmt.Printf("%s\n", addBinary("11", "1"))
	fmt.Printf("%s\n", addBinary("11", "0"))
	fmt.Printf("%s\n", addBinary("11", "11"))
	fmt.Printf("%s\n", addBinary("10", "1"))
}
